### PR TITLE
[Bugfix] Optimize composite weight loading and fix EAGLE weight loading

### PR DESCRIPTION
**Pull request by DarkLight1337**
This PR consolidates and updates the weight loading logic to be more memory efficient, while also automatically detecting child modules and parameters. It also fixes issues with loading EAGLE models due to `lm_head` not existing in the weights.

This should resolve the CI failures related to LoRA and speculative decoding tests.

FIX https://github.com/vllm-project/vllm/pull/9090#issuecomment-2400064662
FIX https://github.com/vllm-project/vllm/pull/9045#issuecomment-2399935631

### CODE REVIEW

**Comment by DarkLight1337** [comment]
@Isotr0py there seems to be a high memory overhead when grouping the weights, any ideas on how to solve this?

**Comment by Isotr0py** [comment]
Maybe it's because of the `itertools.tee`. I'm thinking if we can store the group results in a "lazy" dictionary so that we can prevent the weights repeat, and finish the grouping process within one iteration.

**Comment by Isotr0py** [comment]
> I am thinking of deferring weight loading even during the grouping process, only loading them when the weights in that group are actually loaded.

I think this would be a good way to solve the memory overhead.

**Comment by DarkLight1337** [comment]
> > I am thinking of deferring weight loading even during the grouping process, only loading them when the weights in that group are actually loaded.
> 
> I think this would be a good way to solve the memory overhead.

I have committed a first draft, see if it works. Currently we don't automatically handle custom `load_weights` in modules that are not direct children of the preceding one, but so far there isn't any model that does that yet.

File: vllm/model_executor/models/utils.py [hunk 1]
```diff
@@ -12,55 +12,167 @@
...
+    the weights only once.
+
+    The weight loading logic for individual modules can be overridden
+    by defining a ``load_weights`` method.
+
+    Similarly, the weight loading logic for individual parameters can be
+    overridden by defining a ``weight_loader`` method.
     """
-    init_weights, repeated_weights = itertools.tee(weights, 2)
-    weights_prefix = {name.split(".")[0] for name, _ in init_weights}
```
Comment (Isotr0py) [inline vllm/model_executor/models/utils.py:57]: Ooops, I found here is the root cause of the memory overhead, because this step is not lazy!

**Comment by DarkLight1337** [comment]
OK, this PR is fully ready for review now.

**Comment by gshtras** [comment]
Will test this on MI300 machine now. Repeating the findings from the original PR here for the record:
After further investigation the issue seems to be about the entire model being loaded on the cpu first. And in case of tp>1 it loads the entire unsliced models that many times, so you're likely to hit CPU OOM

**Comment by gshtras** [comment]
Verified `python benchmarks/benchmark_latency.py --model meta-llama/Llama-2-70b-chat-hf -tp 8` on a machine with 1TB RAM, which would previously run out of memory on this configuration, it seems to work now with this fix.

**Comment by DarkLight1337** [comment]
@youkaichao the tests pass now. Can you take a look at this ASAP to unblock the CI failures on main?

**Review by youkaichao — APPROVED** [review]
approve to unblock ci

**Pull request merged**